### PR TITLE
Typed ember glimmer

### LINKED ISSUE #15729: Convert ember-glimmer package to typescript
Convert ember-glimmer to typescript.  Use d.ts files for imports from other packages for now.

### PR DESCRIPTION
**Pull request by krisselden**
Fixes #15729

### CODE REVIEW

**Comment by krisselden** [comment]
Fixes #15729

File: packages/ember-glimmer/lib/environment.ts [hunk 1]
```diff
@@ -203,13 +212,14 @@ export default class Environment extends GlimmerEnvironment {
...
   }
 
-  hasPartial(name: string, { owner }) {
-    return hasPartial(name, owner);
+  hasPartial(partialName: string, meta: any): boolean {
+    return hasPartial(name, meta.owner);
   }
 
-  lookupPartial(name, { owner }) {
+  lookupPartial(PartialName: string, meta: any): PartialDefinition<any> {
```
Comment (smfoote) [inline packages/ember-glimmer/lib/environment.ts]: s/PartialName/name/

File: packages/ember-glimmer/lib/component-managers/outlet.ts [hunk 2]
```diff
@@ -57,38 +57,38 @@ class OutletComponentManager extends AbstractManager<StateBucket> {
     return new StateBucket(outletState);
   }
 
-  layoutFor(definition, bucket, env) {
+  layoutFor(definition, _bucket, env: Environment) {
```
Comment (smfoote) [inline packages/ember-glimmer/lib/component-managers/outlet.ts]: I was wondering how to ignore some parameters.

**Comment by chadhietala** [comment]
Are we now doing the dangling comma?

**Comment by rwjblue** [comment]
DANGLE IT

**Comment by krisselden** [comment]
@chadhietala are we not supposed to? I just did the tslint defaults and turned off the ones that seemed too cumbersome

**Comment by chadhietala** [comment]
I hate it, but if we use prettier I can deal with it.

**Comment by chancancode** [comment]
We should probably use the glimmer rules (or adapt that to fit our actual
preference). Having two sets of rules for those two projects seems
confusing.

**Comment by rwjblue** [comment]
Rebased on top of changes from https://github.com/emberjs/ember.js/pull/15716 (which basically introduced a conflict in all the files 😩 )...

**Comment by chadhietala** [comment]
I am +1 on this.

**Comment by rwjblue** [comment]
🎉

**Comment by rwjblue** [comment]
There is still work needed here, but this seems stable (and is passing CI).

**Pull request merged**